File: frameworks/Kotlin/hexagon/src/main/kotlin/co/there4/hexagon/Benchmark.kt
package co.there4.hexagon

import co.there4.hexagon.serialization.convertToMap
import co.there4.hexagon.serialization.serialize
import co.there4.hexagon.server.Call
import co.there4.hexagon.server.Router
import co.there4.hexagon.server.Server
import co.there4.hexagon.server.jetty.JettyServletEngine
import co.there4.hexagon.server.router
import co.there4.hexagon.server.servlet.ServletServer
import co.there4.hexagon.settings.SettingsManager.settings
import co.there4.hexagon.templates.pebble.PebbleEngine

import java.lang.System.getProperty
import java.lang.System.getenv
import java.util.*
import java.util.concurrent.ThreadLocalRandom
import javax.servlet.annotation.WebListener
import java.net.InetAddress.getByName as address

// DATA CLASSES
internal data class Message(val message: String)
internal data class Fortune(val _id: Int, val message: String)
internal data class World(val _id: Int, val id: Int, val randomNumber: Int)

// CONSTANTS
private const val TEXT_MESSAGE: String = "Hello, World!"
private const val CONTENT_TYPE_JSON = "application/json"
private const val QUERIES_PARAM = "queries"

// UTILITIES
internal fun randomWorld() = ThreadLocalRandom.current().nextInt(WORLD_ROWS) + 1

private fun Call.returnWorlds(worldsList: List<World>) {
    val worlds = worldsList.map { it.convertToMap() - "_id" }
    val result = if (worlds.size == 1) worlds.first().serialize() else worlds.serialize()

    ok(result, CONTENT_TYPE_JSON)
}

private fun Call.getWorldsCount() = (request[QUERIES_PARAM]?.toIntOrNull() ?: 1).let {
    when {
        it < 1 -> 1
        it > 500 -> 500
        else -> it
    }
}

// HANDLERS
private fun Call.listFortunes(store: Store) {
    val fortunes = store.findAllFortunes() + Fortune(0, "Additional fortune added at request time.")
    val locale = Locale.getDefault()
    response.contentType = "text/html; charset=utf-8"
    template(PebbleEngine, "fortunes.html", locale, "fortunes" to fortunes.sortedBy { it.message })
}

private fun Call.getWorlds(store: Store) {
    returnWorlds(store.findWorlds(getWorldsCount()))
}

private fun Call.updateWorlds(store: Store) {
    returnWorlds(store.replaceWorlds(getWorldsCount()))
}

// CONTROLLER
private fun router(): Router = router {
    val store = createStore(getProperty("DBSTORE") ?: getenv("DBSTORE") ?: "mongodb")

    before {
        response.addHeader("Server", "Servlet/3.1")
        response.addHeader("Transfer-Encoding", "chunked")
        response.addHeader("Date", httpDate())
    }

    get("/plaintext") { ok(TEXT_MESSAGE, "text/plain") }
    get("/json") { ok(Message(TEXT_MESSAGE).serialize(), CONTENT_TYPE_JSON) }
    get("/fortunes") { listFortunes(store) }
    get("/db") { getWorlds(store) }
    get("/query") { getWorlds(store) }
    get("/update") { updateWorlds(store) }
}

@WebListener class Web : ServletServer () {
    override fun createRouter() = router()
}

internal var server: Server? = null

fun main(vararg args: String) {
    server = Server(JettyServletEngine(), settings, router()).apply { run() }
}

File: frameworks/PHP/mixphp/src/Controller/Benchmark.php
<?php

namespace App\Controller;

use App\Container\DB;
use Mix\Vega\Context;

class Benchmark
{

    /**
     * @param Context $ctx
     */
    public function json(Context $ctx)
    {
        $ctx->setHeader('Content-Type', 'application/json');
        $ctx->string(200, json_encode(['message' => 'Hello, World!']));
    }

    /**
     * @param Context $ctx
     */
    public function plaintext(Context $ctx)
    {
        $ctx->setHeader('Content-Type', 'text/plain; charset=utf-8');
        $ctx->string(200, 'Hello, World!');
    }

    /**
     * @param Context $ctx
     */
    public function db(Context $ctx)
    {
        $ctx->setHeader('Content-Type', 'application/json');
        $ret = DB::instance()->raw('SELECT id,randomNumber FROM World WHERE id=?', mt_rand(1, 10000))->first();
        $ctx->string(200, json_encode($ret));
    }

    /**
     * @param Context $ctx
     */
    public function query(Context $ctx)
    {
        $queryCount = 1;
        $q = (int)$ctx->query('q');
        if ($q > 1) {
            $queryCount = min($q, 500);
        }

        $arr = [];
        while ($queryCount--) {
            $ret = DB::instance()->raw('SELECT id,randomNumber FROM World WHERE id=?', mt_rand(1, 10000))->first();
            $arr[] = $ret;
        }

        $ctx->setHeader('Content-Type', 'application/json');
        $ctx->string(200, json_encode($arr));
    }

    /**
     * @param Context $ctx
     */
    public function fortunes(Context $ctx)
    {
        $rows = DB::instance()->raw('SELECT id,message FROM Fortune')->get();
        $rows[] = (object)['id' => 0, 'message' => 'Additional fortune added at request time.'];
        usort($rows, function ($left, $right) {
            return $left->message <=> $right->message;
        });
        $ctx->HTML(200, 'fortunes', ['rows' => $rows]);
    }

    /**
     * @param Context $ctx
     */
    public function update(Context $ctx)
    {
        $queryCount = 1;
        $q = (int)$ctx->query('q');
        if ($q > 1) {
            $queryCount = min($q, 500);
        }

        $arr = [];
        while ($queryCount--) {
            $id = mt_rand(1, 10000);
            $ret = DB::instance()->raw('SELECT id,randomNumber FROM World WHERE id=?', $id)->first();
            DB::instance()->exec('UPDATE World SET randomNumber=? WHERE id=?', $ret->randomNumber = mt_rand(1, 10000), $id);
            $arr[] = $ret;
        }

        $ctx->setHeader('Content-Type', 'application/json');
        $ctx->string(200, json_encode($arr));
    }

}
